File: bridge/resources/ios/NotificationRule.swift
import Foundation

/// What the device would do with a notification, read rather than asked.
///
/// The distinction this exists for is between *nobody has been asked* and *the
/// operator said no*. An app that cannot tell them apart prompts a second time
/// on a device where somebody already refused, which is the one thing the rule
/// about a declined permission forbids.
///
/// iOS answers the question directly — `UNAuthorizationStatus` has a
/// `.notDetermined` of its own — so the two states this has to reconstruct on
/// Android arrive here already separated. The rule is written anyway, and takes
/// the same four inputs, because the failure these files exist to catch is the
/// two platforms quietly disagreeing, and a rule that exists on one side only
/// cannot disagree visibly.
public struct NotificationRule: Equatable, Sendable {
    /// Whether a notification posted now would actually appear.
    ///
    /// The whole answer where it is true, and deliberately the first question:
    /// authorisation is not the only switch, and any of them being off means
    /// the same thing to somebody waiting to be told something.
    public let wouldAppear: Bool

    /// Whether this platform has an authorisation to ask for at all.
    ///
    /// Always true on iOS, where notifications have always required one. The
    /// input exists so that the two rules take the same shape; Android's answer
    /// is false below API 33, where silence is a refusal already given rather
    /// than a question still open.
    public let permissionIsAsked: Bool

    /// Whether the platform says an explanation would help.
    ///
    /// Always false on iOS, which offers no equivalent of Android's rationale
    /// flag — and needs none, because `.notDetermined` is answered directly.
    public let wouldExplain: Bool

    /// Whether this application has ever raised the prompt.
    ///
    /// Recorded by this plugin where the prompt is raised. iOS does not need it
    /// to reach the right answer; it is carried so that a disagreement between
    /// the platforms is a disagreement about the same four facts.
    public let everAsked: Bool

    /// The four facts a platform reports, in the order the rule reads them.
    public init(wouldAppear: Bool, permissionIsAsked: Bool, wouldExplain: Bool, everAsked: Bool) {
        self.wouldAppear = wouldAppear
        self.permissionIsAsked = permissionIsAsked
        self.wouldExplain = wouldExplain
        self.everAsked = everAsked
    }

    /// The three answers the application reasons in, as the wire spells them.
    public func said() -> String {
        if wouldAppear { return NotificationRule.granted }
        if !permissionIsAsked { return NotificationRule.denied }
        if wouldExplain { return NotificationRule.denied }
        if everAsked { return NotificationRule.denied }

        return NotificationRule.notDetermined
    }

    /// A notification posted now would appear.
    public static let granted = "granted"

    /// It would not, and nothing may ask again.
    public static let denied = "denied"

    /// Nobody has been asked; the point of first use is still ahead.
    public static let notDetermined = "not_determined"
}

File: bridge/resources/android/NotificationRule.kt
package app.lemonfiber.native

/**
 * Whether a notification would be seen, and whether anybody may still be asked.
 *
 * The four facts a platform can report about notifications, and what they mean
 * together. Reconstructing the three answers out of them is
 * [WhatTheOperatorSaid]'s, because the camera asks the identical question and
 * two copies of that decision would be two capabilities disagreeing quietly.
 * What is here is the part that belongs to telling somebody: which facts are
 * gathered, and what the answer is called once the subject is a notification.
 *
 * No Android framework in sight. Everything is a reading of four booleans a
 * caller passes in, which is what lets it be run on a JVM in two seconds rather
 * than demonstrated on a handset.
 *
 * Deliberately mirrors `NotificationRule.swift` line for line.
 */
public data class NotificationRule(
    /**
     * Whether a notification posted right now would be seen.
     *
     * One fact covering three switches — the runtime permission, the channel,
     * and the application's own toggle in system settings — because the
     * operator turning any of them off means the same thing to a caller with
     * something to show.
     */
    public val wouldAppear: Boolean,
    /**
     * Whether this platform has a runtime permission to ask for at all.
     *
     * False below Android 33, where notifications are granted at install time
     * and turned off in settings afterwards; always true on iOS. Taken as an
     * input rather than branched on inside, so that the version check stays in
     * the shim where the platform is and the decision stays here where it can
     * be run.
     */
    public val permissionIsAsked: Boolean,
    /**
     * Whether the platform says an explanation would help.
     *
     * True only after a refusal, which is what makes it evidence of one rather
     * than a suggestion. It is never true before the first prompt and never
     * true after a permanent one, so it identifies exactly one of the three
     * states and cannot stand alone. Always false on iOS, which has nothing of
     * the kind.
     */
    public val wouldExplain: Boolean,
    /**
     * Whether this application has ever raised the prompt.
     *
     * Kept by this plugin because nothing else can keep it: Android has no
     * call that answers it, and the two states it separates — never asked, and
     * refused so firmly the platform stopped offering — are identical from the
     * outside. Recorded where the prompt is raised rather than inferred
     * afterwards.
     */
    public val everAsked: Boolean,
) {
    /** What the operator has said, as far as anything can tell. */
    public val said: WhatTheOperatorSaid
        get() = WhatTheOperatorSaid.readFrom(wouldAppear, permissionIsAsked, wouldExplain, everAsked)

    /**
     * Whether the prompt may be raised now.
     *
     * Named here as well as on the answer because the shim asks this question
     * of the rule rather than of the word: what a caller has in hand is the
     * four facts, and the two readings of them belong side by side.
     */
    public val mayAsk: Boolean
        get() = said.mayAsk

    /** Whether something posted now would reach the operator. */
    public val mayShow: Boolean
        get() = said.mayProceed

    /** Where the starting state lives, named so it reads at a call site. */
    public companion object {
        /**
         * A device nobody has asked anything, on a platform that asks.
         *
         * What a first launch looks like on Android 33 and above, and the state
         * the other cases are written as a departure from.
         */
        public val UNASKED: NotificationRule =
            NotificationRule(
                wouldAppear = false,
                permissionIsAsked = true,
                wouldExplain = false,
                everAsked = false,
            )
    }
}
